specs/labels-and-localization/README.md

```diff
@@ -0,0 +1,293 @@
+- Grouped indicator in table column headers: Currently no ARIA attributes indicate a grouped column, so we'll probably want to add a label for this (probably via a `title` / tooltip)
+- Client-provided icon elements in the table (e.g. icons as the primary content of table headers): Follows the same guidance in [accessibility.md](../../packages/nimble-components/docs/accessibility.md), i.e. the icons should provide accessible text via the `title` attribute (and clients will handle localizing it themselves).
+- (Mapping/ Icon Columns: As currently specced, `nimble-table-column-icon`'s `nimble-mapping-icon` has a `label` attribute which will become the icon `title`)
+
+## Localization
+
+Our general approach to localization is that clients will handle localizing strings. That has led us to add attributes on some Nimble components for clients to provide localized labels, but that doesn't scale well to components with many labels like the table.
+
+### Plan for Nimble-Provided Labels
+
+Conceptually the localized strings are very similar to the sharing pattern of Design Tokens. The vast majority of the time, you want to use the same value (i.e. numeric increment/decrement button text is not control instance specific). However, we do want to be able to override the strings for specific controls as needed.
+
+We will create non-CSS-property design tokens, similar to the `theme` and `direction` design tokens.
+We will also create new `nimble-i18n-*` elements with APIs for setting the tokens. These will generally be direct children of the `nimble-theme-provider`, and the theme provider will be the target that the token values are set on.
+```html
+<nimble-theme-provider>
+  <nimble-i18n-core
+    number-field-increment="Increment"
+    number-field-decrement="Decrement"
+  ></nimble-i18n-core>
+  <nimble-i18n-table
+    table-groups-collapse-all="Collapse all groups"
+  ></nimble-i18n-table>
+</nimble-theme-provider>
+```
+Pros:
+- Can configure once for the whole page in one spot, does not need to be per control. Much fewer duplicated attributes / slotted elements in the page.
+- Can override for specific elements as needed (by wrapping in `nimble-theme-provider` and `nimble-i18n-*` elements)
+- We can make more granular i18n providers, even down to the level of individual controls, if we're concerned about the number of strings apps would be automatically pulling in. (Note that currently we're just planning to have 2 though.)
+- Code outside of Nimble could use the same concept (and derive from our i18n base class). If we end up with new controls in Nimble that are mostly specific to specific apps, we can also create new i18n providers for those elements, rather than sticking them all in the `i18n-core` element.
```

Comment:
I forget, where did we end up on client libraries defining their own elements? Would systemlink-lib-angular expose its own or would it just document that apps should pull in one or both of Nimble's (depending on which components they use)?

For ease of implementation I'd be fine if we started out with only the 2 Nimble elements. Almost every SLE app will pull them both in so there's not much initial benefit to sl-lib trying to be smarter about it.

Reply:
Current expectation is that client apps would just pull in these (`nimble-i18n-core and potentially nimble-i18n-table`). The only expected scenario where there'd be additional i18n providers would be if there were extensions to Nimble outside Nimble that wanted to use the same system.

I decided to prototype pulling these changes into TestInsights just to validate all of my assumptions.
Branch/commit: [users/masmith/nimble-localization-prototyping](https://dev.azure.com/ni/DevCentral/_git/Skyline/commit/346747fa80736d43be39f559d58d8958f6757d5c?refName=refs/heads/users/masmith/nimble-localization-prototyping)

Everything mostly worked like I expected, but 2 surprises:
- As soon as apps depend on the nimble-angular version with these changes, running `i18n:extract` for that app yields the Nimble labels in the app messages.xlf files for localization, even if the app doesn't yet depend on `NimbleInternationalizationCoreModule`/`NimbleInternationalizationTableModule`. I think that's still OK though since we expect apps to do both updates at the same time (pull in new version, add `nimble-i18n-core/table withDefaults` in their `nimble-theme-provider`, add imports to the modules in their app module).
    - It's possible that secondary entry points for the i18n stuff in nimble-angular would change that behavior some, but I haven't prototyped that so far.
- The Nimble strings may get merged into the same translation units as the existing strings in SystemLink, if the strings match. For example my prototype has banner-dismiss="Close" and that gets merged:
```xml
<trans-unit id="7819314041543176992" datatype="html">
	<source>Close</source>
	<target>Schließen</target>
	<context-group purpose="location">
		<context context-type="sourcefile">node_modules/projects/file-manager-lib/src/files/upload-files/upload-files/upload-files.component.html</context>
		<context context-type="linenumber">1</context>
	</context-group>
	<context-group purpose="location">
		<context context-type="sourcefile">projects/ni/nimble-angular/src/directives/i18n/core/nimble-i18n-core-with-defaults.directive.ts</context>
		<context context-type="linenumber">15</context>
	</context-group>
	<context-group purpose="location">
		<context context-type="sourcefile">src/app/dataspaces/components/axis-selection-slideout/axis-selection-slideout.component.ts</context>
		<context context-type="linenumber">14</context>
	</context-group>
	<context-group purpose="location">
		<context context-type="sourcefile">src/app/dataspaces/components/color-by-slideout/color-by-slideout.component.ts</context>
		<context context-type="linenumber">13</context>
	</context-group>
	<context-group purpose="location">
		<context context-type="sourcefile">src/app/dataspaces/components/dataspaces-query-builder/dataspaces-query-builder.component.ts</context>
		<context context-type="linenumber">23</context>
	</context-group>
	<context-group purpose="location">
		<context context-type="sourcefile">src/app/results/components/step-details-slide-out/step-details-slide-out.component.ts</context>
		<context context-type="linenumber">21</context>
	</context-group>
</trans-unit>
```
I even tried adding a description on the Nimble side (`` $localize`:Nimble banner dismiss button:Close`; ``) but got the same behavior.

Reply:
Updated the wording in the Angular section to state that pulling in the nimble-angular version and running extract-i18n will result in the Nimble strings getting pulled in (rather than importing the modules being the act that does it).

Also added a sentence stating that we expect most SLE apps to consume both i18n-core and i18n-table as-is.

Reply:
Just as one more followup, I updated the prototypes to create secondary entry points for each i18n provider, and verified that when nimble-angular was set up that way, apps will only pull in our strings once they import the i18n providers specifically from those entry points. So that's probably what we want to do (updated the HLD).